### PR TITLE
[C API] Add blockaddress getters to C API

### PR DESCRIPTION
**Pull request by Benjins**
Exposes getters for the function and basic block associated with blockaddress constants to the C API. Tests for them are added, which required a tweak to llvm-c-test to declare basic blocks ahead of any actual function cloning, since the blockaddress constants may reference basic blocks in arbitrary other functions

----------------------------

This PR is split into two commits for reviewing sake. The first commit is the refactor of `echo.cpp` to pre-declare all basic blocks in `declare_symbols` along with the functions. The second commit contains the pertinent change to the C API

### CODE REVIEW

**Comment by Benjins** [comment]
Bumping this for review

File: llvm/tools/llvm-c-test/echo.cpp [hunk 1]
```diff
@@ -373,6 +387,22 @@ static LLVMValueRef clone_constant_impl(LLVMValueRef Cst, LLVMModuleRef M) {
...
 
+  if (LLVMIsABlockAddress(Cst)) {
+    check_value_kind(Cst, LLVMBlockAddressValueKind);
+    LLVMValueRef SrcFunc = LLVMGetBlockAddressFunction(Cst);
+    LLVMBasicBlockRef SrcBB = LLVMGetBlockAddressBasicBlock(Cst);
+
+    LLVMValueRef DstFunc = clone_constant(SrcFunc, M);
+
+    LLVMBasicBlockRef DstBB =
+        find_bb_in_func(DstFunc, LLVMGetBasicBlockName(SrcBB));
```
Comment (nikic) [inline llvm/tools/llvm-c-test/echo.cpp:398]: This approach is not going to work for unnamed blocks.

**Comment by Benjins** [comment]
I have pushed some changes to deal with unnamed basic blocks. From my understanding, `blockaddress` constants can only reference unnamed basic blocks that are in the same function. So to support that case, we clone them inside the `FunCloner` (which has access to the `BBMap` needed to look up unnamed basic blocks), or outside functions in which case we can only clone them by name as the previous version of the PR did. This also means that we only pre-declare named basic blocks, and unnamed ones are lazily cloned in `DeclareBB` as before

I added a couple more tests, for unnamed local basic blocks, and a global `blockaddress` constant. I decided to move these to a dedicated test file since they were getting to be more substantial

**Comment by nikic** [comment]
> I have pushed some changes to deal with unnamed basic blocks. From my understanding, `blockaddress` constants can only reference unnamed basic blocks that are in the same function.

It looks like for textual IR, we allow it if the reference comes first (https://llvm.godbolt.org/z/8K7rq418Y), but forbid it when it comes later (https://llvm.godbolt.org/z/dj4n7b8zb). Presumably so we don't have to keep around the per-function-state on the off-chance of a later blockaddress reference.

> So to support that case, we clone them inside the `FunCloner` (which has access to the `BBMap` needed to look up unnamed basic blocks), or outside functions in which case we can only clone them by name as the previous version of the PR did. This also means that we only pre-declare named basic blocks, and unnamed ones are lazily cloned in `DeclareBB` as before
> 
> I added a couple more tests, for unnamed local basic blocks, and a global `blockaddress` constant. I decided to move these to a dedicated test file since they were getting to be more substantial

Hm, I'm wondering whether it wouldn't make more sense to simply add a unit to test for the blockaddress getters, instead of supporting them inside echo? Blockaddress cloning is tricky, and I'm not sure there's a lot of value trying to make echo support it correctly, at the cost of extra complexity.

**Comment by Benjins** [comment]
I made a new PR with the same API changes, but with a simpler unit test for them (modeling off other non-echo test functionality in `llvm-c-test`). Will close this one since it's now superseded
https://github.com/llvm/llvm-project/pull/81382

**Pull request closed without merging**